xla/stream_executor/cuda/cuda_dnn.cc

```diff
@@ -4675,21 +4824,782 @@ GetCudnnFusedMHAOperationGraph(
+  TF_ASSIGN_OR_RETURN(
+      auto tensor_ds,
+      CreateCudnnSoftmaxBwdTensor(intermdiate_ops, p_dims, p_strides, dtype,
+                                  tensor_p_abs, tensor_dp_scale_dropout));
+  // mask backward
+  TF_ASSIGN_OR_RETURN(
+      auto tensor_ds_mask,
+      CreateCudnnMaskBwdTensor(intermdiate_ops, p_dims, p_strides, dtype,
+                               tensor_ds, use_mask));
+
+#if (CUDNN_VERSION >= 8901 && TF_ENABLE_CUDNN_FRONTEND)
+  // bias backward
+  TF_ASSIGN_OR_RETURN(auto tensor_dbias, CreateCudnnBiasBwdTensor(
+                                             intermdiate_ops, p_dims, p_strides,
+                                             dtype, tensor_ds_mask));
+#else
+  return absl::InternalError("Bias backward op requires cudnn >= 8.9.1");
 #endif
 
+  // calculate dq
+  auto bmm1_grad_gemm2_desc = cudnn_frontend::MatMulDescBuilder()
+                                  .setComputeType(CUDNN_DATA_FLOAT)
+                                  .build();
+  RETURN_MSG_IF_CUDNN_ERROR(bmm1_grad_gemm2_desc);
+  auto bmm1_grad_gemm2_op = cudnn_frontend::OperationBuilder(
+                                CUDNN_BACKEND_OPERATION_MATMUL_DESCRIPTOR)
+                                .setaMatDesc(tensor_ds_mask)
+                                .setbMatDesc(tensor_k)
+                                .setcMatDesc(tensor_dq)
+                                .setmatmulDesc(bmm1_grad_gemm2_desc)
+                                .build();
+  RETURN_MSG_IF_CUDNN_ERROR(bmm1_grad_gemm2_op);
+  intermdiate_ops.push_back(std::move(bmm1_grad_gemm2_op));
+
+  // calculate dk
+  TF_ASSIGN_OR_RETURN(auto tensor_ds_mask_reshape,
+                      CreateCudnnTensor(p_transpose_dims, p_transpose_strides,
+                                        VIRTUAL_ID + 305, dtype, 1, -1,
+                                        /* is_virtual */ true));
+
+  auto reshape_2_op = cudnn_frontend::OperationBuilder(
+                          CUDNN_BACKEND_OPERATION_RESHAPE_DESCRIPTOR)
+                          .setxDesc(tensor_ds_mask)
+                          .setyDesc(tensor_ds_mask_reshape)
+                          .build();
+
+  intermdiate_ops.push_back(std::move(reshape_2_op));
+  auto bmm1_grad_gemm1_desc = cudnn_frontend::MatMulDescBuilder()
+                                  .setComputeType(CUDNN_DATA_FLOAT)
+                                  .build();
+  RETURN_MSG_IF_CUDNN_ERROR(bmm1_grad_gemm1_desc);
+  auto bmm1_grad_gemm1_op = cudnn_frontend::OperationBuilder(
+                                CUDNN_BACKEND_OPERATION_MATMUL_DESCRIPTOR)
+                                .setaMatDesc(tensor_ds_mask_reshape)
+                                .setbMatDesc(tensor_q)
+                                .setcMatDesc(tensor_dk)
+                                .setmatmulDesc(bmm1_grad_gemm1_desc)
+                                .build();
+  RETURN_MSG_IF_CUDNN_ERROR(bmm1_grad_gemm1_op);
+  intermdiate_ops.push_back(std::move(bmm1_grad_gemm1_op));
+
```

Comment:
Possible to do `ops.reserve` here?

Reply:
added reserve, also changed push_back to emplace_back